### PR TITLE
Feature: BB-309 add oplog populator metrics

### PR DESCRIPTION
**Pull request by welansari**
Issue: [BB-309](https://scality.atlassian.net/browse/BB-309)

[BB-309]: https://scality.atlassian.net/browse/BB-309?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[BB-309]: https://scality.atlassian.net/browse/BB-309?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: extensions/oplogPopulator/OplogPopulator.js [hunk 1]
```diff
@@ -67,6 +76,7 @@ class OplogPopulator {
             const client = await MongoClient.connect(this._mongoUrl, {
                  replicaSet: this._replicaSet,
                  useNewUrlParser: true,
+                 useUnifiedTopology: true,
```
Comment (francoisferrand) [inline extensions/oplogPopulator/OplogPopulator.js:79]: seems unrelated...

Comment (welansari) [reply extensions/oplogPopulator/OplogPopulator.js:79]: yes the two first commits are drive by fixes of some issues i found

File: extensions/oplogPopulator/OplogPopulator.js [hunk 2]
```diff
@@ -6,12 +6,15 @@ const { constructConnectionString } = require('../utils/MongoUtils');
...
 const ConnectorsManager = require('./modules/ConnectorsManager');
+const { ZenkoMetrics } = require('arsenal').metrics;
+const OplogPopulatorMetrics = require('./OplogPopulatorMetrics');
 
 const paramsJoi = joi.object({
     config: joi.object().required(),
     mongoConfig: joi.object().required(),
     activeExtensions: joi.array().required(),
     logger: joi.object().required(),
+    enableMetrics: joi.boolean().default(true),
```
Comment (francoisferrand) [inline extensions/oplogPopulator/OplogPopulator.js:17]: do we need to make metrics optional?
i don't really see a use-case here or "precedent" in other modules, shouldn't metrics always be collected?

File: extensions/oplogPopulator/OplogPopulator.js (hunk 1 above)
Comment (francoisferrand) [reply extensions/oplogPopulator/OplogPopulator.js:79]: we may want to validate these drive-by fixes in zenko CI before merging though...

File: extensions/oplogPopulator/OplogPopulator.js [hunk 3]
```diff
@@ -200,7 +212,16 @@ class OplogPopulator {
                     '_id': 1,
                     'operationType': 1,
                     'documentKey._id': 1,
-                    'fullDocument.value': 1
+                    'fullDocument.value': 1,
+                    // transforming the BSON timestamp
+                    // into a usable date
```
Comment (francoisferrand) [inline extensions/oplogPopulator/OplogPopulator.js:217]: clusterTime should already be included, with a value like this, right?
```js
"clusterTime": Timestamp({ "t": 1666193824, "i": 1 }),
```

can't just read the timestamp (`clusterTime.t`) in oplogpopulator instead of converting to date string and parsing the date?

File: extensions/oplogPopulator/OplogPopulatorMetrics.js [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        this.requestSize = null;
+        this.connectors = null;
+        this.reconfigurationLag = null;
+        this._logger = logger;
+    }
+
+    registerMetrics() {
+        this.acknowledgementLag = ZenkoMetrics.createHistogram({
+            name: 'oplog_populator_acknowledgement_lag_sec',
+            help: 'Delta between a config change in mongo and it getting processed by the oplogPopulator in seconds',
```
Comment (francoisferrand) [inline extensions/oplogPopulator/OplogPopulatorMetrics.js]: _**Delay**_ between a config change in mongo and _**start of processing**_ by the oplogPopulator in seconds

File: extensions/oplogPopulator/OplogPopulatorMetrics.js [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+            labelNames: ['opType'],
+            buckets: [0.001, 0.01, 1, 10, 100, 1000, 10000],
+        });
+        this.connectorConfiguration = ZenkoMetrics.createCounter({
```
Comment (francoisferrand) [inline extensions/oplogPopulator/OplogPopulatorMetrics.js:26]: is it the number of "reconfiguration" (e.g. running the whole loop)?
or the number of time we update the configuration of a single connector?

File: extensions/oplogPopulator/OplogPopulatorMetrics.js [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+            labelNames: ['connector'],
+        });
+        this.connectors = ZenkoMetrics.createGauge({
+            name: 'oplog_populator_connectors',
+            help: 'Total number of configured connectors',
+            labelNames: ['isOld'],
+        });
+        this.reconfigurationLag = ZenkoMetrics.createHistogram({
+            name: 'oplog_populator_reconfiguration_lag_sec',
+            help: 'Time it takes to reconfigure a connector',
```
Comment (francoisferrand) [inline extensions/oplogPopulator/OplogPopulatorMetrics.js]: is it the time it takes to configure a _single_ connector? (e.g. issuing the command to kafka connect)
or does it also include the time it takes for a _single_ connector to "be ready" (e.g. tasks re-created/updated in kafka connect)?
or the overall "reconfigure time" (that it takes to reconfigure all connectors)

File: extensions/oplogPopulator/OplogPopulatorMetrics.js [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+            labelNames: ['connector'],
+            buckets: [0.001, 0.01, 1, 10, 100, 1000, 10000],
+        });
+        this.reconfiguration = ZenkoMetrics.createCounter({
```
Comment (francoisferrand) [inline extensions/oplogPopulator/OplogPopulatorMetrics.js]: not clear what this measures & what "unit"

File: extensions/oplogPopulator/OplogPopulator.js (hunk 2 above)
Comment (welansari) [reply extensions/oplogPopulator/OplogPopulator.js:17]: In normal cases we always have the metrics enabled, that's why it's set to true by default.
The option was put there to be able to run unit tests with no issues, as otherwise we would get errors about the metrics being registered multiple times.

File: extensions/oplogPopulator/OplogPopulator.js (hunk 3 above)
Comment (welansari) [reply extensions/oplogPopulator/OplogPopulator.js:217]: I did this because i find it cleaner than the way we do in Arsenal which is not precise, i'd rather let mongo transform it into a usable timestamp.
```
 timestamp: new Date((itemObj.ts ? itemObj.ts.getHighBits() * 1000 : 0))
```

Comment (welansari) [reply extensions/oplogPopulator/OplogPopulator.js:217]: And the value `clusterTime` is equal to:
```
Timestamp { _bsontype: 'Timestamp', low_: 1, high_: 1672913394 }
```

File: extensions/oplogPopulator/OplogPopulator.js (hunk 1 above)
Comment (welansari) [reply extensions/oplogPopulator/OplogPopulator.js:79]: No need i think since the code only impacts the oplog populator (no tests for now), and i tested everything locally and on an Artesca cluster

File: extensions/oplogPopulator/OplogPopulator.js (hunk 2 above)
Comment (francoisferrand) [reply extensions/oplogPopulator/OplogPopulator.js:17]: we may indeed have something for the tests, but it should not propagate to the config...

**Review by francoisferrand — APPROVED** [review]

File: extensions/oplogPopulator/modules/ConnectorsManager.js [hunk 8]
```diff
@@ -204,12 +208,14 @@ class ConnectorsManager {
                 const oldConnectors = await this._getOldConnectors(oldConnectorNames);
                 this._connectors.push(...oldConnectors);
                 this._oldConnectors.push(...oldConnectors);
+                this._metricsHandler.onConnectorsInstantiated(true, oldConnectors.length);
```
Comment (XinLi0207) [inline extensions/oplogPopulator/modules/ConnectorsManager.js:211]: onConnectorsInstantiated() catches the error inside the function if `this.connectors.inc` failes, but doesn't it matter for the following logic of  `this._metricsHandler.onConnectorsInstantiated(true, oldConnectors.length);` even if `this.connectors.inc` fails?

Comment (welansari) [reply extensions/oplogPopulator/modules/ConnectorsManager.js:211]: It does affect it, the value of the metric will be false.
We don't however want to crash the oplog populator here as that will lead to some events being missed in case one or more bucket configurations change.

**Review by XinLi0207 — APPROVED** [review]
